src/System.Linq.Parallel/tests/ExchangeTests.cs

```diff
@@ -10,183 +11,177 @@ namespace Test
 {
     public class ExchangeTests
     {
-        [Fact]
-        public static void SimplePartitionMergeWhereScanTest()
+        private static readonly ParallelMergeOptions[] Options = new[] {
+            ParallelMergeOptions.AutoBuffered,
+            ParallelMergeOptions.Default,
+            ParallelMergeOptions.FullyBuffered,
+            ParallelMergeOptions.NotBuffered
+        };
+
+        public static IEnumerable<object[]> PartitioningData(object[] counts)
         {
-            for (int i = 1; i <= 128; i *= 2)
-            {
-                SimplePartitionMergeWhereScanTest1(1024 * 8, i, true);
-                SimplePartitionMergeWhereScanTest1(1024 * 8, i, false);
-            }
+            foreach (object[] results in Sources.Ranges(counts.Cast<int>(), x => new[] { 1, 2, 4 })) yield return results;
         }
 
-        [Fact]
-        [OuterLoop]
-        public static void SimplePartitionMergeWhereScanTest_LongRunning()
+        public static IEnumerable<object[]> MergeData(object[] counts)
         {
-            for (int i = 1; i <= 128; i *= 2)
-            {
-                SimplePartitionMergeWhereScanTest1(1024 * 1024 * 2, i, true);
-                SimplePartitionMergeWhereScanTest1(1024 * 1024 * 2, i, false);
-            }
+            foreach (object[] results in Sources.Ranges(counts.Cast<int>(), x => Options)) yield return results;
         }
 
-        private static void SimplePartitionMergeWhereScanTest1(int dataSize, int partitions, bool pipeline)
+        public static IEnumerable<object[]> FailingMergeData(object[] counts)
         {
-            int[] data = new int[dataSize];
-            for (int i = 0; i < dataSize; i++)
-                data[i] = i;
-
-            var whereOp = data.AsParallel()
-                .Where(delegate (int x) { return (x % 2) == 0; }); // select only even elements
-
-            IEnumerator<int> stream = whereOp.GetEnumerator();
-
-            int count = 0;
-            while (stream.MoveNext())
+            foreach (int count in counts.Cast<int>())
             {
-                // @TODO: verify all the elements we expected are present.
-                count++;
+                Labeled<ParallelQuery<int>> labeled = Labeled.Label("ThrowOnEnumeration " + count, new ThrowOnEnumeration<int>(count).AsParallel().AsOrdered());
+                foreach (ParallelMergeOptions option in Options)
+                {
+                    yield return new object[] { labeled, count, option };
+                }
             }
+        }
+
+        // The following tests attempt to test internal behavior,
+        // but there doesn't appear to be any way to deterministically observe it.
```

Comment:
I was trying to say "We have tests for internal behavior, but there isn't a reliable way to externally observe it" (The tests already existed, I just translated/widened them.  I otherwise left them in their "current" state). 

 Take the "partitioning" case.  From the DegreesOfParallelism tests, we know PLINQ will use multiple threads (.... eventually, when large).  What we don't have a direct path to observe is the number of internal partitions it creates and uses, at least when using the current types directly (array, enumerable, etc).

I guess I could write that custom partitioner, for these tests?  The feeling I got was that degrees of parallelism wasn't necessarily directly related to the partitions created. 
(Although I'm realizing you're right, and I should create a bunch of custom implementations for various error conditions, like null enumerables out of `GetDynamicPartitions`).